tests/CommonExtractor.php

```diff
@@ -0,0 +1,497 @@
+
+        if (is_array($resultRow) && !empty($resultRow)) {
+            // write header and first line
+            if ($includeHeader) {
+                $csvWriter->writeRow(array_keys($resultRow));
+            }
+            $csvWriter->writeRow($resultRow);
+
+            // write the rest
+            $numRows = 1;
+            $lastRow = $resultRow;
+
+            while ($resultRow = $stmt->fetch(\PDO::FETCH_ASSOC)) {
+                $csvWriter->writeRow($resultRow);
+                $lastRow = $resultRow;
+                $numRows++;
+            }
+            $stmt->closeCursor();
+
+            if (isset($this->incrementalFetching['column'])) {
+                if (!array_key_exists($this->incrementalFetching['column'], $lastRow)) {
+                    throw new UserException(
+                        sprintf(
+                            "The specified incremental fetching column %s not found in the table",
+                            $this->incrementalFetching['column']
+                        )
+                    );
+                }
+                $output['lastFetchedRow'] = $lastRow[$this->incrementalFetching['column']];
+            }
+            $output['rows'] = $numRows;
+            return $output;
+        }
+        // no rows found.  If incremental fetching is turned on, we need to preserve the last state
+        if ($this->incrementalFetching['column'] && isset($this->state['lastFetchedRow'])) {
+            $output = $this->state;
+        }
+        $output['rows'] = 0;
+        return $output;
+    }
+
+    private function createConnection(DatabaseParameters $parameters): \PDO
+    {
+        if ($parameters->getPort()) {
+            $dsn = sprintf(
+                "mysql:host=%s;port=%s;dbname=%s;charset=utf8",
+                $parameters->getHost(),
+                $parameters->getPort(),
+                $parameters->getDatabase()
+            );
+        } else {
+            $dsn = sprintf(
+                "mysql:host=%s;dbname=%s;charset=utf8",
+                $parameters->getHost(),
+                $parameters->getDatabase()
+            );
+        }
+        $options = [
+            \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION,
+        ];
+        try {
+            return new \PDO($dsn, $parameters->getUser(), $parameters->getPassword(), $options);
+        } catch (\Throwable $exception) {
+            throw new UserException("Error connecting to DB: " . $exception->getMessage(), 0, $exception);
+        }
+    }
+
+    private function executeQuery(string $query, ?int $maxTries): \PDOStatement
+    {
+        $proxy = new RetryProxy($this->getLogger(), $maxTries);
+        $stmt = $proxy->call(function () use ($query) {
+            try {
+                /** @var \PDOStatement $stmt */
+                $stmt = $this->getConnection()->prepare($query);
+                $stmt->execute();
+                return $stmt;
+            } catch (\Throwable $e) {
+                try {
```

Comment:
proč tady je tohle? od toho, když to failne je přece ta RetryProxy? Tak proč se tady s tím dělají takové tanečky?

Reply:
(ptám se protože podle toho ostatní asi budou dělat svoje extractory)

Reply:
protoze retryproxy jen znovu spusti tu query, ale nenavaze spojeni, cili pokusi se to tady navazat spojeni, ale kdyz se to nepovede (coz se muze snadno stat), tak chces delat retry na te originalni vyjimce, ne na te zpusobene tim reconnectem.
bylo by to elegantnejsi, kdyby ta proxy mela onRetry callback - to myslim resil @pivnicek

Reply:
jo, mam ten retryProxy v devu tady https://github.com/keboola/db-extractor-common/pull/78, bude tam ten callback, dobry napad!